File: Lab3.py
compare = lambda s1, s2: s1<s2 if max(s1)==max(s2) else max(s1)<max(s2) 

            

def quicksort(x):
    if len(x) == 1 or len(x) == 0:
        return x

    else:
        pivot = x[0]
        i = 0
        for j in range(len(x)-1):
            if  compare(x[j+1],pivot):
                x[j+1],x[i+1] = x[i+1], x[j+1]
                i += 1

        x[0],x[i] = x[i],x[0] #swap

        first_part = quicksort(x[:i])
        second_part = quicksort(x[i+1:])
        first_part.append(x[i])
        return first_part + second_part


arr = raw_input().split()

arr = quicksort(arr)

sort = lambda s1,s2: max(s1) < max(s2) 


for i in arr:
    print i,

File: Lab3.java
import java.util.*; 
import java.io.*; 

public class Lab4{
	public static void main (String args[]) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in)); 

		int size = 	Integer.parseInt(br.readLine()); 
		
		String arr[] = new String[size];

		for(int i = 0; i<size; i++)
			arr[i] = br.readLine(); 

		quickSort(arr,0,size-1); 


		for(String s: arr)
			System.out.print(s +" "); 
		System.out.println(); 

	}

	public static void swap(int i, int j,String arr[]){
		String temp = arr[i];
		arr[i] = arr[j];
		arr[j] = temp; 
	}


	public static char greatestCharacter(String s){
		char g = s.charAt(0);

		for(int i = 1; i<s.length(); i++)
			if(s.charAt(i) > g)
				g = s.charAt(i);

		return g;
	}

	public static boolean compare(String s1, String s2){
		char g1 = greatestCharacter(s1);
		char g2 = greatestCharacter(s2); 

		if(g1==g2){
			if(s2.compareTo(s1)>0)
				return true;
			
			return false;
		}

		else return g1<=g2; 
	}

  public static void quickSort(String arr[], int low, int high){
    String pivot = arr[low + (high-low)/2];
    int i = low;
    int j = high;

    while(i<=j){

      while(compare(arr[i],pivot)){
        i++;
      }

      while(compare(pivot,arr[j])){
        j--;
      }

      if(i<=j){
        swap(i,j,arr);
        i++;
        j--;
        //print();
      }
    }

    if(low < j){
      quickSort(arr, low, j);
    }

    if(high>i){
      quickSort(arr, i, high);
    }
  }
} 
